lib/client.js

```diff
@@ -934,23 +1081,49 @@ MqttClient.prototype._handlePublish = function (packet, done) {
   var that = this
 
   switch (qos) {
-    case 2:
-      this.incomingStore.put(packet, function () {
-        that._sendPacket({cmd: 'pubrec', messageId: mid}, done)
-      })
+    case 2: {
+      if (this.options.customHandleAcks) {
+        this.options.customHandleAcks(topic, message, packet, function (code) {
+          if (code) {
+            that._sendPacket({cmd: 'pubrec', messageId: mid, reasonCode: code}, done)
+          } else {
+            that.incomingStore.put(packet, function () {
+              that._sendPacket({cmd: 'pubrec', messageId: mid}, done)
+            })
+          }
+        })
+      } else {
+        this.incomingStore.put(packet, function () {
+          that._sendPacket({cmd: 'pubrec', messageId: mid}, done)
+        })
+      }
       break
-    case 1:
+    }
+    case 1: {
       // emit the message event
-      this.emit('message', topic, message, packet)
-      this.handleMessage(packet, function (err) {
-        if (err) {
-          return done && done(err)
-        }
-        // send 'puback' if the above 'handleMessage' method executed
-        // successfully.
-        that._sendPacket({cmd: 'puback', messageId: mid}, done)
-      })
+      if (this.options.customHandleAcks) {
+        this.options.customHandleAcks(topic, message, packet, function (code) {
```

Comment:
the signature for the callback of `customHandleAcks` must be `function (err, code) {}`. We should also validate that `code` is valid, or is this done when generating a PUBACK`?

Reply:
code validation added